File: python/4-median-of-two-sorted-arrays.py
class Solution:

    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # 设置两个索引, index1, index2分别跟随nums1, nums2
        # first,second 代表中间值和中间值前面的那个值
        first = second = 0
        index1 = index2 = 0
        len1 = len(nums1)
        len2= len(nums2)
        mid = (len1+len2) //2 +1
        for i in range(mid):
            second = first
            if( (index2 >= len2) or ((index1 < len1) and (nums1[index1] <= nums2[index2]))):
                first = nums1[index1]
                index1 += 1
            else:
                first = nums2[index2]
                index2 += 1
        
        if((len1+len2) % 2 == 0):
            return (first+second)/2
        else:
            return float(first)

''' #快速解法
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        nums = sorted(nums1 + nums2)
        mid = len(nums)//2
        if(len(nums)%2 == 0):
            return (nums[mid] + nums[mid-1])/2
        else:
            return float(nums[mid])
'''

File: java/4-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        //分别为两个数组的索引
        int index1 = 0, index2 = 0;
        //因为不知道两个数组长度之和为奇为偶,所以中位数要保存两个
        int fast = 0, slow = 0;
        //找中位数只需要遍历一般就行了
        int lens = nums1.length + nums2.length;
        int mid = lens / 2 + 1;
        for(int i = 0; i < mid; i++){
            //更新slow的值
            slow = fast;
            if( (index2 >= nums2.length)  || ( (index1 < nums1.length) && (nums1[index1] <= nums2[index2]))) {
                fast = nums1[index1];
                index1++;
            }else{
                fast = nums2[index2];
                index2++;
            }
        }
        if(lens % 2 == 0){
            return (fast+slow)/2.0;
        }else{
            return fast*1.0;
        }
    }
}
